### PR TITLE
New Meshline

### PR DESCRIPTION
**Pull request by Mugen87**
This PR introduces a new and flexible way of drawing lines via [Screen-Space Projected Lines](https://mattdesl.svbtle.com/drawing-lines-is-hard#screenspace-projected-lines_2). The implementation is based on https://github.com/spite/THREE.MeshLine by @spite.

[**Demo**](https://yume.human-interactive.org/examples/meshline/webgl_meshline.html)

You can create a mesh line like this:

```javascript
var line = new THREE.MeshLine( geometry, material );
```
- `geometry` can be a `Geometry` or `BufferGeometry`. Users can provide vertices (position) and color data (for vertex colors).
- `material` must be a `MeshBasicMaterial`. It's the only material that is supported right now. You can use three new material properties to control the visual appearance of the line: `meshLine`(like `morphTargets`) , `meshLineWidth`, `meshLineSizeAttenuation`. Other material properties like `color`, `transparent` or `opacity` also work.
- Besides, there is a third and optional parameter of `MeshLine` that can be used to create nice curve shapes (`widthCallback`).

As soon as we commit to a final API design and naming conventions, i will also create some docs with a subsequent PR.

There is still an open issue. I'm not sure how to handle the situation if the geometry of a mesh line changes. `MeshLine` always needs to process the given geometry to prepare the data for the vertex shader. The original [MeshLine]( https://github.com/spite/THREE.MeshLine) project does this with a separate method `setGeometry`. But i wonder if we can find a way that fits better to the library conventions and is maybe more convenient.

### CODE REVIEW

**Comment by mrdoob** [comment]
You're such a life saver! ❤️

**Comment by mrdoob** [comment]
What do you think about make this transparent though? So if `THREE.Line`'s `linewidth > 1` then `WebGLRenderer` uses this code. Otherwise it goes through the usual path.

We'll have to lose `widthCallback` but, at the end of the day, I think it would be great if `THREE.Line` `linewidth` just worked.

**Comment by Mugen87** [comment]
>  So if THREE.Line's linewidth > 1 then WebGLRenderer uses this code. Otherwise it goes through the usual path.

After some research i was not able to find a nice integration of both approaches. But to be honest, i don't know all parts of `WebGLRenderer` in detail. If there is a smart solution to do this, i'm of course open to it. But currently it seems to me, that the code in  `WebGLRenderer` will be more complicated if we mix the rendering of line primitives and screen-space projected lines.

Therefore i would propose something different :sweat_smile:. 

`Line`, `LineLoop` and `LineSegments` should be exclusively used for rendering line primitives. Because the usage of `gl.lineWidth()` is not reliable ([reference](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/lineWidth)), i would remove `linewidth` from `LineBasicMaterial` and `LineDashedMaterial` and render all line primitives with the default line width of `1`. If users still apply `linewidth > 1` to a `Line*` material, `three.js` could log a warning to use `MeshLine` in combination with `MeshBasicMaterial` instead.

**Comment by fernandojsg** [comment]
> So if THREE.Line's linewidth > 1 then WebGLRenderer uses this code. Otherwise it goes through the usual path.

As this is screen space implementation it won't work properly when using WebVR. That's why I needed to create a actual geometry mesh line for [a-painter](https://aframe.io/a-painter/).
I'd planned to send a PR to add it, we could probably use three of them:
- If `non-vr` and `Linewidth == 1` -> `THREE.line`
- If `non-vr` and `LineWidth > 1` -> `THREE.MeshLineSS`
- If `vr` -> `THREE.MeshLine`.

Actually for me is also confusing the name `MeshLine` as I expected it to be real geometry mesh instead of a SS representation, so probably renaming it to a more clear name could be nice too.
Another problem I got with Jaume`s `THREE.MeshLine` implementation is that it lacks of most of the material options that we have for the rest of geometry and that could be applied to an actual geometry mesh line.

File: examples/webgl_meshline.html [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+				camera.lookAt( scene.position );
+
+				var time = Date.now() * 0.0005;
+
+				for ( var i = 0; i < scene.children.length; i ++ ) {
+
+					var object = scene.children[ i ];
+
+					if ( object instanceof THREE.Line ) {
```
Comment (WestLangley) [inline examples/webgl_meshline.html]: `THREE.MeshLine`

File: examples/webgl_meshline.html [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+						object.rotation.y = time * ( i % 2 ? 1 : -1 );
+
+					}
+
+				}
+
+				// animate line width
+
+				material1.meshLineWidth = 10 * ( 1 + 0.5 * Math.sin( 5 * time ) );
```
Comment (WestLangley) [inline examples/webgl_meshline.html]: material2.meshLineWidth = 100 * ( 1 + 0.5 * Math.sin( 5 * time ) );

Suggestion. Vary `material2` instead, and use `OrbitControls`. It is easier to zoom and see the subtitles of what is going on.

File: examples/webgl_meshline.html (hunk 1 above)
Comment (Mugen87) [inline examples/webgl_meshline.html]: :+1:

File: examples/webgl_meshline.html (hunk 2 above)
Comment (Mugen87) [inline examples/webgl_meshline.html]: Good idea!

**Comment by Mugen87** [comment]
I'm not sure about the way how to handle different line implementations. We could create separate entities like `THREE.Line` or `THREE.MeshLine` and let the user decide what to use. On the other hand we can have just a single entity (e.g. `THREE.Line`) and run transparently different implementations dependent on various factors (material properties or VR/non-VR). Currently, i tend to favor the first approach. 

@fernandojsg, @WestLangley What's your opinion about this?

**Comment by Mugen87** [comment]
> That's why I needed to create a actual geometry mesh line for a-painter.

@fernandojsg Can you share the code for this? I am interested to see your solution :blush:

File: examples/webgl_meshline.html (hunk 2 above)
Comment (WestLangley) [inline examples/webgl_meshline.html]: `100 *`
please so it is visible on small devices.

Comment (WestLangley) [inline examples/webgl_meshline.html]: Perhaps also set size attenuation on the first one.

What are the units of `width`?

**Comment by WestLangley** [comment]
I am not sure `MeshLine` is a good name for a screen-space approach. Trying to think of an alternate nomenclature...

- If non-vr and Linewidth == 1 -> `THREE.Line`
- If non-vr and LineWidth > 1 -> `THREE.ThickLine`  ??? or `THREE.Line` if implementation is transparent
- If vr -> `THREE.Ribbon` ???

File: examples/webgl_meshline.html (hunk 2 above)
Comment (Mugen87) [inline examples/webgl_meshline.html]: Sry, overlooked the 100* thing. I've also added size attenuation to the first material.

Comment (Mugen87) [inline examples/webgl_meshline.html]: 'meshLineWidth' should be equal to the value of `gl.lineWidth()` so it's pixel...

Comment (WestLangley) [inline examples/webgl_meshline.html]: Thanks for this. You can change it later if you want. For now, it is helpful. :)

**Comment by mrdoob** [comment]
>> That's why I needed to create a actual geometry mesh line for a-painter.
>
> Can you share the code for this? I am interested to see your solution :blush:

He's probably creating a tube. Which makes me think that we could also create a tube too..

**Comment by WestLangley** [comment]
>What are the units of width?
>>meshLineWidth should be equal to the value of gl.lineWidth() so it's pixel...

If it is consistent with THREE.Points, then fine. If not, we should figure out why. Also consider DPR.

**Comment by fernandojsg** [comment]
Thinking about names I like @WestLangley proposal for `THREE.Ribbon` , as it's actually that, not a tube. 
I'm not sure about the approach discussed for vr/non-vr and linewidths as they're not really consistent for example regarding materials, so probably I'll just leave it to the user to choose which one they want to use.

I'll try to open a PR for the `THREE.Ribbon` later this week.

**Comment by WestLangley** [comment]
We [used to have](https://github.com/mrdoob/three.js/issues/2523#issuecomment-9529672) `THREE.Ribbon`. I do not recall the implementation details, though. It was removed for some reason...

**Comment by spite** [comment]
It's not a Ribbon, either, if we're being technical.

I'm afraid you're trying to cram way too functionality in Line: WebGL Lines are mathematical lines, and there's the unfortunate problem with lineWidth across platforms, which makes sense, since the definition of a line is precisely that doesn't have breadth or thickness. 

That's the main and only thing to solve for a Line replacement (and it could be built into the THREE.Line family transparently). Other stuff, like adding properties from other materials or rendering lines in VR (what does that even mean!), it should just be a Mesh, any Mesh: CurvePath, ShapePath, Ribbon, what have you.

Just my two cents.

**Comment by mrdoob** [comment]
Agreed. I think we should leave the VR use case out of this discussion. We don't want to turn `THREE.Line` into TiltBrush. We just want to have lines with lineWidth > 1, which is something that used to work (OSX only anyway).

**Comment by mrdoob** [comment]
> After some research i was not able to find a nice integration of both approaches. But to be honest, i don't know all parts of `WebGLRenderer` in detail. If there is a smart solution to do this, i'm of course open to it. But currently it seems to me, that the code in  `WebGLRenderer` will be more complicated if we mix the rendering of line primitives and screen-space projected lines.

So, unless someone else tries it before me, I'll use this code as reference and try to add support for `lineWidth` > 1 in `WebGLRenderer` for r86.

**Comment by Mugen87** [comment]
Sounds good to me. If i can manage it,  i'll have a look at the integration of the code into `WebGLRenderer`  in the next days again. For the case i make any kind of progress, i'll post it here.

**Comment by fernandojsg** [comment]
Ok I agree, it sound good to me too, better to leave the "`ribbon`" out from this and just in case add as an external example but nothing more.
Maybe just for the `vr` case, we could just force use simple lines instead of the SS ones.

**Comment by Mugen87** [comment]
> Also consider DPR.

@WestLangley That was a good hint! :+1:  The last commit makes the line width of `MeshLine` now consistent to `THREE.Line`.

**Comment by Mugen87** [comment]
Related issues: #10686, #10357

**Comment by WestLangley** [comment]
When the number of points is reduced, there appears to be a problem with the implementation.
```
var spline = new THREE.CatmullRomCurve3( points );
var divisions = points.length / 6; // previously points.length * 6
```
screen shot 2017-05-10 at 2 33 07 pm

**Comment by Mugen87** [comment]
I think a too small sampling rate causes issues when calculating the line extrusions in the vertex shader. This seems to be a conceptual problem of screen-space projected lines.

**Comment by Mugen87** [comment]
see #11349

**Pull request closed without merging**